Issue #49: Fixed bug where FindParameter does not consider repeat parameters

hexthedev opened the pull request:
See Issue for bug details: https://github.com/criteo/openapi-comparator/issues/49

TLDR: Basically, even though our spec doesn't change, there are breaking changes detected when two parameters have the same name, but differ in values.

To solve this, instead of using the `OpenApiParameter.name` value to find the first parameter with a matching name in the new spec, I collect all parameters with the name and do a comparison. 

The comparison compares what I considered functional values (i.e. values that change the way the route acts). So not descriptions. But things like the `in` value. A score is generated based on the number of these things that are alike, then it chooses the one with the highest score.

It is possible that two things have equal alikeness, in this case the first one will be picked. This is likely an indicator of an actual breaking change since a parameter would have changed functionally. So it is okay that this edge case is present.

ouvreboite commented on src/Criteo.OpenApi.Comparator/Comparators/OperationComparator.cs:
```diff
@@ -257,30 +257,75 @@ private static int FindParameterIndex(OpenApiParameter parameter, IList<OpenApiP
+                return null;
+
+            if(candidateParameters.Count == 0)
+                return candidateParameters[0];
```
That will throw an exception. I think you meant:
```
if(candidateParameters.Count == 1)
  return candidateParameters[0];
```

hexthedev replied:
Good catch, I'll add a unity test that causes this failure then fix it.

hexthedev replied:
Just realized that 

```
if(candidateParameters.Count == 1)
  return candidateParameters[0];
```

Never actually happened because of the Any check above. So it was skipped and `GetClosestFunctionally` was just returning the 1 candidate any way.

ouvreboite commented on src/Criteo.OpenApi.Comparator/Comparators/OperationComparator.cs:
```diff
@@ -257,30 +257,75 @@ private static int FindParameterIndex(OpenApiParameter parameter, IList<OpenApiP
+                score += key.AllowEmptyValue == candidate.AllowEmptyValue ? 1 : 0;
+                score += key.Style == candidate.Style ? 1 : 0;
+                score += key.Explode == candidate.Explode ? 1 : 0;
+                score += key.AllowReserved == candidate.AllowReserved ? 1 : 0;
```
That's an interesting take, but I worry that it would make the comparison less predictable.
Could you add some tests to showcase a bit what the behavior could be?

Also, I feel this would miss the most important similarity for path parameters: the order in which they are present in the path.

hexthedev replied:
I was thinking this also. I'm wondering if a better approach might be the index of the parameter in the path, but I think the order of header parameters and query parameters doesn't actually matter. Also if an optional parameter is added, it will move a parameter. So I thought that parameter index wouldn't be as useful in the end as values. 

This comparison is a best guess on my part, so I'm open to better comparison methods. My actual intention is to find the parameter that EXACTLY matches the old parameter, no matter where it appears in the parameters list. Then, if that does not exist, take a reasonable remaining candidate. This will maintain the behaviour that catch legitimate `1015`s, without falsely identifying some. 

Anyway, I'll write a few more tests to make sure that there is a decent rail to prove it works properly.

ouvreboite reviewed (changes requested):
Hi 👋, thank you for this PR. I've left a few comments below.

hexthedev commented on src/Criteo.OpenApi.Comparator.UTest/Resource/issue_49_false_1015_equal_schemas/diff.json:
```diff
@@ -0,0 +1,13 @@
+[
```
This test shows that when two parameters are present with the same name, #49 does not occur. This is the equal case, where the schemas do not differ.

hexthedev commented on src/Criteo.OpenApi.Comparator.UTest/Resource/issue_49_false_1015_out_of_order_parameters/diff.json:
```diff
@@ -0,0 +1,35 @@
+[
```
This test shows that when two parameters and out of order, 1042 issues are still detected. This is the equal case, where the schemas differ in parameter order.

hexthedev commented:
@ouvreboite : I've added one more simple test. Are there any other tests you would like to see?

I added tests that cover the problem. I am relying on existing tests to indicate that I did not cause other parts of the package to fail. I thought about adding 'normal' situations, but these are all covered by the existing suite. 

My hope is that this covers an edge case, while keeping the normal case simple. The score based comparison should only occur in the case where two parameters have the same name. I think most of the time, they'll simply match on `candidate.In`

I could simply do that comparison that way. Just check if a candidate matches on the `In` parameter. I think my OpenAPI knowledge isn't enough to make the call.

ouvreboite reviewed (approved):
LGTM
